cpp/src/io/parquet/rle_stream.cuh

```diff
@@ -56,92 +55,102 @@ inline __device__ uint32_t get_vlq32(uint8_t const*& cur, uint8_t const* end)
   return v;
 }
 
-// an individual batch. processed by a warp.
-// batches should be in shared memory.
-template <typename level_t>
-struct rle_batch {
-  uint8_t const* run_start;  // start of the run we are part of
-  int run_offset;            // value offset of this batch from the start of the run
-  level_t* output;
-  int level_run;
-  int size;
-
-  __device__ inline void decode(uint8_t const* const end, int level_bits, int lane, int warp_id)
-  {
-    int output_pos = 0;
-    int remain     = size;
+template <typename level_t, int max_output_values>
+__device__ inline void decode(level_t* const output,
+                              uint8_t const* const run_start,
+                              int const run_offset,
+                              int const run_output_pos,
+                              int const level_run,
+                              int const size,
+                              uint8_t const* const end,
+                              int level_bits,
+                              int lane)
+{
+  // local output_pos for this `decode` call.
+  int decode_output_pos = 0;
+  int remain            = size;
 
-    // for bitpacked/literal runs, total size is always a multiple of 8. so we need to take care if
-    // we are not starting/ending exactly on a run boundary
-    uint8_t const* cur;
-    if (level_run & 1) {
-      int const effective_offset = cudf::util::round_down_safe(run_offset, 8);
-      int const lead_values      = (run_offset - effective_offset);
-      output_pos -= lead_values;
-      remain += lead_values;
-      cur = run_start + ((effective_offset >> 3) * level_bits);
-    }
+  // for bitpacked/literal runs, total size is always a multiple of 8. so we need to take care if
+  // we are not starting/ending exactly on a run boundary
+  uint8_t const* cur;
+  if (level_run & 1) {
```

Comment:
Part of me wants to add a couple of functions. ` is_repeated_run()` and` is_literal_run()`. It's just so easy to confuse the two.

Reply:
Done https://github.com/rapidsai/cudf/pull/15159/commits/a17cdf91f03b533c15cc7b98679c1455def8a02e